src/storage/secondary/column.rs

```diff
@@ -149,42 +149,41 @@ impl Column {
-            // No need to verify checksum when read from cache
-            do_verify_checksum = false;
-            block
-        } else {
-            // block has not been in cache, so we fetch it from disk
-            // TODO(chi): support multiple I/O backend
-
-            let file = self.file.clone();
-            let info = self.index.index(block_id).clone();
-            let block = tokio::task::spawn_blocking(move || {
-                let mut data = vec![0; info.length as usize];
-                // TODO(chi): handle file system errors
-                match file {
-                    ColumnReadableFile::PositionedRead(file) => {
-                        file.read_exact_at(&mut data[..], info.offset)?;
+        let mut do_verify_checksum = false;
+
+        // support multiple I/O backend
+        let block = self
+            .block_cache
+            .get_or_try_insert_with(key, async {
+                // block has not been in cache, so we fetch it from disk
+                let file = self.file.clone();
+                let info = self.index.index(block_id).clone();
+                let block = tokio::task::spawn_blocking(move || {
+                    let mut data = vec![0; info.length as usize];
+                    // TODO(chi): handle file system errors
+                    match file {
+                        ColumnReadableFile::PositionedRead(file) => {
+                            file.read_exact_at(&mut data[..], info.offset)?;
+                        }
+                        ColumnReadableFile::NormalRead(file) => {
+                            let mut file = file.lock().unwrap();
+                            file.seek(SeekFrom::Start(info.offset as u64))?;
+                            file.read_exact(&mut data[..])?;
+                        }
                     }
-                    ColumnReadableFile::NormalRead(file) => {
-                        let mut file = file.lock().unwrap();
-                        file.seek(SeekFrom::Start(info.offset as u64))?;
-                        file.read_exact(&mut data[..])?;
-                    }
-                }
-                Ok::<_, TracedStorageError>(Bytes::from(data))
+                    Ok::<_, TracedStorageError>(Bytes::from(data))
+                })
+                .await
+                .unwrap();
+                // TODO(chi): we should invalidate cache item after a RowSet has been compacted.
+                // self.block_cache.insert(key, block.clone()).await;
+
+                // need to verify checksum when read from disk
+                do_verify_checksum = true;
+                block
             })
             .await
-            .unwrap()?;
-
-            // TODO(chi): we should invalidate cache item after a RowSet has been compacted.
-            self.block_cache.insert(key, block.clone()).await;
-
-            do_verify_checksum = true;
-            block
-        };
+            .unwrap();
```

Comment:
Is there anyway to return this error? e.g. add a `Arc<StroageError>` variant for `StorageError`?

Reply:
i'm sorry, i not very clear about "add a `Arc<StroageError>` variant for `StorageError`", i tried to use `?` replace `unwrap`, but this need impl `From` trait(I'm not particularly familiar with this part)

It's all I can think of to do
```rust
let block = match block {
    Ok(block) => block,
    Err(_) => return Err(TracedStorageError::not_found("block", block_id)),
};
```
Maybe I need some examples...[rust rookie :)]

Reply:
https://github.com/risinglightdb/risinglight/blob/733be2953af4b434ccd0c59f40aa71f638d4887d/src/storage/error.rs#L9-L33

Here we can add a new nested enum variant:

```rust
    #[error("{0}")]
    #[backtrace]
    #[from]
    Nested(Arc< StorageError>),
}
```

So that we can directly use `?` on such error.

Reply:
and the `get_or_try_insert_with` will be panics when the `init` future has been panicked, Whether a file system error causes panic?

Reply:
ok, i will to fix it after i finish this PR, hhh!

Reply:
Hey, I went to look at the code and I think it should return error instead of panic, the file system error will pass to `block` of `spawn_blocking` and pass to `block` of `get_or_try_insert_with`.
Of course, if a file system panic occurs, the task calling this method will panic, other task will continue work.

Reply:
Agree. You may send a new PR, thanks!

Reply:
> Hey, I went to look at the code and I think it should return error instead of panic, the file system error will pass to `block` of `spawn_blocking` and pass to `block` of `get_or_try_insert_with`. Of course, if a file system panic occurs, the task calling this method will panic, other task will continue work.

Sorry, I may have a problem with my expression (forgive my broken English), I mean it can do that now. Don't to change any code, just remove the `TODO` comment :)

Reply:
Sorry, I misunderstood you. A new PR to remove `TODO` is welcomed!